shared/mad/codeql/mad/dynamic/GraphExport.qll

```diff
@@ -0,0 +1,230 @@
+   */
+  bindingset[host]
+  predicate hasTypeSummary(Node host, string path);
+
+  /**
+   * Holds if paths going through `node` should be blocked.
+   *
+   * For example, this can be the case for functions that are public at runtime
+   * but intended to be private.
+   */
+  predicate shouldNotContain(Node node);
+}
+
+/**
+ * Module for exporting an arbitrary graph as models-as-data rows.
+ */
+module GraphExport<NodeSig Node, GraphExportSig<Node> S> {
+  private import S
+
+  private Node getAnExposedNode() {
+    not shouldNotContain(result) and
+    (
+      exposedName(result, _, _)
+      or
+      edge(getAnExposedNode(), _, result)
+    )
+  }
+
+  pragma[nomagic]
+  private predicate exposedEdge(Node pred, string path, Node succ) {
+    // Materialize this relation so we can access 'edge' without binding set on 'pred'
+    pred = getAnExposedNode() and
+    edge(pred, path, succ)
+  }
+
+  private Node getARelevantNode() {
+    result = getAnExposedNode() and
+    shouldContain(result)
+    or
+    exposedEdge(result, _, getARelevantNode())
+  }
+
+  final private class FinalNode = Node;
+
+  private class RelevantNode extends FinalNode {
+    RelevantNode() { this = getARelevantNode() }
+  }
+
+  pragma[inline]
+  private RelevantNode getAPredecessor(RelevantNode node, string path) {
+    exposedEdge(result, path, node)
+  }
+
+  private predicate nodeMustBeNamed(RelevantNode node) {
+    exposedName(node, _, "")
+    or
+    S::mustBeNamed(node)
+    or
+    strictcount(getAPredecessor(node, _)) > 1
+  }
+
+  /** Gets a type-name to use as a prefix, in case we need to synthesize a name. */
+  private string getAPrefixTypeName(RelevantNode node) {
+    result =
+      min(string prefix |
+        exists(string type, string path |
+          exposedName(node, type, path) and
+          prefix = join(type, path)
+        )
+        or
+        suggestedName(node, prefix)
+      |
+        prefix
+      )
+    or
+    not exposedName(node, _, _) and
+    not suggestedName(node, _) and
+    result = getAPrefixTypeName(getAPredecessor(node, _))
+  }
+
+  /**
+   * Holds if a named type exists or will be generated for `node`.
+   */
+  private predicate isSyntheticallyNamedNode(RelevantNode node, string prefix) {
+    nodeMustBeNamed(node) and
+    not exposedName(node, _, "") and
+    not suggestedName(node, _) and
+    prefix = min(getAPrefixTypeName(node))
+  }
+
+  /**
+   * Gets a synthetic type name to generate for `node`.
+   */
+  private string getSyntheticName(RelevantNode node) {
+    exists(int k, string prefixTypeName |
+      node =
+        rank[k](RelevantNode n, string path, int startline, int startcol, int endline, int endcol |
+          isSyntheticallyNamedNode(n, prefixTypeName) and
+          n.hasLocationInfo(path, startline, startcol, endline, endcol)
+        |
+          // Use location information for an arbitrary ordering
+          n order by path, startline, startcol, endline, endcol
```

Comment:
I think because we only use def-nodes and never use-nodes, they can't have the same location. And if they do, it means we spuriously merge two nodes in the graph. But since the nodes occur at the same location it seems unlikely to be a problem, it might even be beneficial in practice.

Reply:
This ties into the other thread above where I suggested naming nodes based on its access path. If the location-clash becomes an issue, I think that's the route we should take. But I'd like to merge it as-is for now.